File: smsly-core/smsly_core/health.py
"""
Unified Health Check Module
===========================
Provides comprehensive health checks with component status for all services.
"""

import time
from typing import Optional, Dict, Any, Callable, Awaitable
from fastapi import APIRouter, Response
from pydantic import BaseModel
from enum import Enum
import structlog

logger = structlog.get_logger(__name__)


class HealthStatus(str, Enum):
    HEALTHY = "healthy"
    DEGRADED = "degraded"
    UNHEALTHY = "unhealthy"


class ComponentHealth(BaseModel):
    status: str
    latency_ms: Optional[float] = None
    error: Optional[str] = None


class HealthResponse(BaseModel):
    status: HealthStatus
    service: str
    version: str
    components: Dict[str, ComponentHealth]
    timestamp: float


async def check_database(engine) -> ComponentHealth:
    """Check database connectivity and latency."""
    try:
        start = time.time()
        async with engine.connect() as conn:
            await conn.execute("SELECT 1")
        latency = (time.time() - start) * 1000
        return ComponentHealth(status="connected", latency_ms=round(latency, 2))
    except Exception as e:
        logger.error("Database health check failed", error=str(e))
        return ComponentHealth(status="error", error=str(e))


async def check_redis(redis_client) -> ComponentHealth:
    """Check Redis connectivity and latency."""
    try:
        start = time.time()
        await redis_client.ping()
        latency = (time.time() - start) * 1000
        return ComponentHealth(status="connected", latency_ms=round(latency, 2))
    except Exception as e:
        logger.error("Redis health check failed", error=str(e))
        return ComponentHealth(status="error", error=str(e))


async def check_rabbitmq(rabbitmq_client) -> ComponentHealth:
    """Check RabbitMQ connectivity."""
    try:
        start = time.time()
        # Check if connection is open
        if hasattr(rabbitmq_client, 'is_closed') and not rabbitmq_client.is_closed:
            latency = (time.time() - start) * 1000
            return ComponentHealth(status="connected", latency_ms=round(latency, 2))
        elif hasattr(rabbitmq_client, 'channel') and rabbitmq_client.channel:
            latency = (time.time() - start) * 1000
            return ComponentHealth(status="connected", latency_ms=round(latency, 2))
        return ComponentHealth(status="disconnected")
    except Exception as e:
        logger.error("RabbitMQ health check failed", error=str(e))
        return ComponentHealth(status="error", error=str(e))


def create_health_router(
    service_name: str,
    version: str = "1.0.0",
    engine=None,
    redis_client=None,
    rabbitmq_client=None,
    custom_checks: Optional[Dict[str, Callable[[], Awaitable[ComponentHealth]]]] = None,
) -> APIRouter:
    """
    Create a health check router with comprehensive component status.
    
    Args:
        service_name: Name of the service (e.g., "smsly-sms")
        version: Service version
        engine: SQLAlchemy async engine (optional)
        redis_client: Redis client (optional)
        rabbitmq_client: RabbitMQ connection (optional)
        custom_checks: Dict of custom health check functions (optional)
    
    Returns:
        FastAPI router with /health, /health/live, and /health/ready endpoints
    """
    router = APIRouter(tags=["Health"])
    
    @router.get("/health", response_model=HealthResponse)
    async def health_check() -> HealthResponse:
        """Comprehensive health check with all component statuses."""
        components: Dict[str, ComponentHealth] = {}
        overall_status = HealthStatus.HEALTHY
        
        # Check database
        if engine is not None:
            db_health = await check_database(engine)
            components["database"] = db_health
            if db_health.status == "error":
                overall_status = HealthStatus.UNHEALTHY
        
        # Check Redis
        if redis_client is not None:
            redis_health = await check_redis(redis_client)
            components["redis"] = redis_health
            if redis_health.status == "error":
                if overall_status == HealthStatus.HEALTHY:
                    overall_status = HealthStatus.DEGRADED
        
        # Check RabbitMQ
        if rabbitmq_client is not None:
            rmq_health = await check_rabbitmq(rabbitmq_client)
            components["rabbitmq"] = rmq_health
            if rmq_health.status == "error":
                if overall_status == HealthStatus.HEALTHY:
                    overall_status = HealthStatus.DEGRADED
        
        # Run custom checks
        if custom_checks:
            for name, check_fn in custom_checks.items():
                try:
                    components[name] = await check_fn()
                except Exception as e:
                    components[name] = ComponentHealth(status="error", error=str(e))
        
        return HealthResponse(
            status=overall_status,
            service=service_name,
            version=version,
            components=components,
            timestamp=time.time(),
        )
    
    @router.get("/health/live")
    async def liveness_probe():
        """Kubernetes liveness probe - always returns 200 if service is running."""
        return {"status": "alive"}
    
    @router.get("/health/ready")
    async def readiness_probe():
        """Kubernetes readiness probe - checks if service is ready to receive traffic."""
        # For readiness, we check critical components only (database)
        if engine is not None:
            db_health = await check_database(engine)
            if db_health.status == "error":
                return Response(
                    content='{"status": "not_ready", "reason": "database_unavailable"}',
                    status_code=503,
                    media_type="application/json",
                )
        return {"status": "ready"}
    
    return router

File: smsly-rust/smsly-core/src/health.rs
use axum::{
    extract::State,
    http::StatusCode,
    response::{IntoResponse, Response},
    routing::get,
    Json, Router,
};
use redis::Client;
use serde::{Deserialize, Serialize};
use sqlx::PgPool;
use std::collections::HashMap;
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};
use tracing::error;

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum HealthStatus {
    Healthy,
    Degraded,
    Unhealthy,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ComponentHealth {
    pub status: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub latency_ms: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct HealthResponse {
    pub status: HealthStatus,
    pub service: String,
    pub version: String,
    pub components: HashMap<String, ComponentHealth>,
    pub timestamp: f64,
}

#[derive(Clone)]
pub struct HealthState {
    pub service_name: String,
    pub version: String,
    pub db_pool: Option<PgPool>,
    pub redis_client: Option<Client>,
}

async fn check_database(pool: &PgPool) -> ComponentHealth {
    let start = SystemTime::now();
    match sqlx::query("SELECT 1").execute(pool).await {
        Ok(_) => {
            let duration = start.elapsed().unwrap_or_default().as_secs_f64() * 1000.0;
            ComponentHealth {
                status: "connected".to_string(),
                latency_ms: Some((duration * 100.0).round() / 100.0),
                error: None,
            }
        }
        Err(e) => {
            error!("Database health check failed: {}", e);
            ComponentHealth {
                status: "error".to_string(),
                latency_ms: None,
                error: Some(e.to_string()),
            }
        }
    }
}

async fn check_redis(client: &Client) -> ComponentHealth {
    let start = SystemTime::now();
    match client.get_multiplexed_async_connection().await {
        Ok(mut conn) => match redis::cmd("PING").query_async::<_, String>(&mut conn).await {
            Ok(_) => {
                let duration = start.elapsed().unwrap_or_default().as_secs_f64() * 1000.0;
                ComponentHealth {
                    status: "connected".to_string(),
                    latency_ms: Some((duration * 100.0).round() / 100.0),
                    error: None,
                }
            }
            Err(e) => {
                error!("Redis PING failed: {}", e);
                ComponentHealth {
                    status: "error".to_string(),
                    latency_ms: None,
                    error: Some(e.to_string()),
                }
            }
        },
        Err(e) => {
            error!("Redis connection failed: {}", e);
            ComponentHealth {
                status: "error".to_string(),
                latency_ms: None,
                error: Some(e.to_string()),
            }
        }
    }
}

pub fn create_health_router(
    service_name: String,
    version: String,
    db_pool: Option<PgPool>,
    redis_client: Option<Client>,
) -> Router {
    let state = HealthState {
        service_name,
        version,
        db_pool,
        redis_client,
    };

    Router::new()
        .route("/health", get(health_handler))
        .route("/health/live", get(liveness_probe))
        .route("/health/ready", get(readiness_probe))
        .with_state(Arc::new(state))
}

async fn health_handler(State(state): State<Arc<HealthState>>) -> Json<HealthResponse> {
    let mut components = HashMap::new();
    let mut overall_status = HealthStatus::Healthy;

    if let Some(pool) = &state.db_pool {
        let h = check_database(pool).await;
        if h.status == "error" {
            overall_status = HealthStatus::Unhealthy;
        }
        components.insert("database".to_string(), h);
    }

    if let Some(client) = &state.redis_client {
        let h = check_redis(client).await;
        if h.status == "error" && overall_status == HealthStatus::Healthy {
            overall_status = HealthStatus::Degraded;
        }
        components.insert("redis".to_string(), h);
    }

    let timestamp = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs_f64();

    Json(HealthResponse {
        status: overall_status,
        service: state.service_name.clone(),
        version: state.version.clone(),
        components,
        timestamp,
    })
}

async fn liveness_probe() -> Json<serde_json::Value> {
    Json(serde_json::json!({"status": "alive"}))
}

async fn readiness_probe(State(state): State<Arc<HealthState>>) -> Response {
    if let Some(pool) = &state.db_pool {
        let h = check_database(pool).await;
        if h.status == "error" {
            return (
                StatusCode::SERVICE_UNAVAILABLE,
                Json(serde_json::json!({"status": "not_ready", "reason": "database_unavailable"})),
            )
                .into_response();
        }
    }
    (StatusCode::OK, Json(serde_json::json!({"status": "ready"}))).into_response()
}
